qutip/solver/ode/explicit_rk.pyx

```diff
@@ -0,0 +1,485 @@
+            self._y_temp = self._accumulate(self._y_temp, self.a[i,:], dt, i)
+            self.k[i] = self.qevo.matmul_data(self._t_prev + self.c[i]*dt,
+                                              self._y_temp, <Data> self.k[i])
+
+    cdef Data _interpolate_step(self, double t, Data out):
+        """
+        Compute the state at any points between _t_prev and _t_front
+        with an error of ~dt**denseout_order.
+        """
+        cdef:
+            int i, j
+            double t0 = self._t_prev
+            double dt = self._dt_int
+            double tau = (t - t0) / dt
+        for i in range(self.rk_extra_step):
+            self.b_factor[i] = 0.
+            for j in range(self.denseout_order-1, -1, -1):
+                self.b_factor[i] += self.bi[i,j]
+                self.b_factor[i] *= tau
+
+        out = copy_to(self._y_prev, out)
+        out = self._accumulate(out, self.b_factor, dt, self.rk_extra_step)
+        return out
+
+    cdef inline Data _accumulate(self, Data target, double[:] factors,
+                         double dt, int size):
+        cdef int i
+        for i in range(size):
+            target = iadd_data(target, <Data> self.k[i], dt * factors[i])
+        return target
+
+    @cython.cdivision(True)
+    cdef double _get_timestep(self, double t):
+        """ Get the dt for the step. """
+        cdef double dt_needed = t - self._t_front
+        if not self.adaptative_step:
+            return dt_needed
+        if self.interpolate:
+            return self._dt_safe
+        if dt_needed <= self._dt_safe:
+            return dt_needed
+        return dt_needed / (int(dt_needed / self._dt_safe) + 1)
+
+    cdef double _recompute_safe_step(self, double err, double dt):
+        """ Get maximum safe step in function of the error."""
+        cdef double factor
+        if not self.adaptative_step:
+            return 0.
+        elif err == 0:
+            factor = 10
+        else:
+            factor = 0.9*err**(-1/(self.order+1))
+            factor = min(10, factor)
+            factor = max(0.2, factor)
+
+        self._dt_safe = dt * factor
+        if self.max_step:
+            self._dt_safe = min(self.max_step, self._dt_safe)
+        if self.min_step:
+            self._dt_safe = max(self.min_step, self._dt_safe)
+
+    def successful(self):
+        return self._status >= 0
+
+    @property
+    def status(self):
+        return self._status
+
+    @property
+    def y(self):
+        return self._y
+
+    @property
+    def y_prev(self):
+        return self._y_prev
+
+    @property
+    def y_front(self):
+        return self._y_front
+
+    @property
+    def t_front(self):
+        return self._t_front
+
+    @property
+    def t_prev(self):
+        return self._t_prev
+
+    @property
+    def t(self):
+        return self._t
+
+    def _debug_state(self):
+        if self._y is not None:
+            print("y", self._t, self._y.to_array())
```

Comment:
For these could we always print `y`, etc but print `None` if the value is None? E.g. `print("y @ t =", self._t, ": ", None if self._y is None else self._y.to_array())`.